twilight-gateway/src/ratelimiter.rs

```diff
@@ -1,127 +1,233 @@
+            let new_deadline = Instant::now() + self.next_available();
+            if self.delay.deadline() < new_deadline {
+                self.delay.as_mut().reset(new_deadline);
+            }
+            ready!(self.delay.as_mut().poll(cx));
+        }
+        Poll::Ready(())
+    }
 
-    // Guard against the interval being 0, in which case we can default.
-    if heartbeat_interval == 0 {
-        return ALLOT_ON_FAIL;
+    /// Cleans up elapsed instants.
+    fn clean(&mut self) {
+        self.instants
+            .retain(|instant| instant.elapsed() < RESET_DURATION);
     }
+}
 
-    let mut heartbeats = RESET_DURATION_MILLISECONDS / heartbeat_interval;
-    let remainder = RESET_DURATION_MILLISECONDS % heartbeat_interval;
+/// Ratelimit permit.
+///
+/// Holding one of these means there's capacity for sending *one* event.
+pub(crate) struct Permit<'a> {
+    /// `true` if the permit was unused.
+    unused: bool,
+    /// Reference to the ratelimiter.
+    inner: &'a mut CommandRatelimiter,
+}
 
-    // If we have a remainder then we reserve an additional heartbeat.
-    //
-    // If there is a remainder per reset then in theory we could allot one less
-    // command for heartbeating variably every number of resets, but it's best
-    // to be cautious and keep it simple.
-    if remainder > 0 {
-        heartbeats = heartbeats.saturating_add(1);
+impl<'a> Permit<'a> {
+    /// Create a new ratelimit permit.
+    fn new(ratelimiter: &'a mut CommandRatelimiter) -> Self {
+        Self {
+            unused: false,
+            inner: ratelimiter,
+        }
     }
 
-    // Convert the heartbeats to a u8. The number of heartbeats **should** never
-    // be above `u8::MAX`, so the error pattern branch should never be reached.
-    let heartbeats_converted = if let Ok(value) = heartbeats.try_into() {
-        value
-    } else {
-        tracing::warn!(
-            default=ALLOT_ON_FAIL,
-            %heartbeats,
-            "heartbeats > u8 max; defaulting",
-        );
+    /// Forget the permit, returning it to the [`CommandRatelimiter`].
+    ///
+    /// Use this when no event was sent.
+    pub(crate) fn forget(mut self) {
+        self.unused = true;
+    }
+}
 
-        ALLOT_ON_FAIL
-    };
+impl Drop for Permit<'_> {
+    fn drop(&mut self) {
+        if !self.unused {
+            self.inner.instants.push(Instant::now());
+        }
+    }
+}
 
-    COMMANDS_PER_RESET.saturating_sub(heartbeats_converted * 2)
+/// Calculate the number of non reserved commands for heartbeating (which skips
+/// the ratelimiter) in a given [`RESET_DURATION`].
+///
+/// Reserves capacity for the amount of heartbeats + 1, to account for Discord
+/// absurdly sending [`OpCode::Heartbeat`]s when the gateway is ratelimited
+/// (which requires the gateway to immediately send a heartbeat back).
+///
+/// [`OpCode::Heartbeat`]: twilight_model::gateway::OpCode::Heartbeat
+fn nonreserved_commands_per_reset(heartbeat_interval: Duration) -> u8 {
```

Comment:
It did not need to be renamed, but I found the "available_commands_per_interval" name to be confusing as the actual available commands per interval is 120 whereas the function actually calculated 120 minus the reserved amount of commands for heartbeating.

Using `as_secs_f32` significantly reduces the LOC which is almost always better for readability. Having to conditionally add a remainder by using of an if statement vs calling `ceil` is incredibly confusing (the only reason for doing so would be an aversion to not following clippy). The clippy allows are trivially redundant: durations are always positive so the cast will never cause a sign loss, and there will be no truncation as the float is always a whole number (`ceil()` rounds up to the nearest whole number).
We might disagree here, with me not wishing to follow clippy if it leads to less readable code, so I guess I'll fold as you're the lead dev but please do confirm.

The old behavior of reserving double the amount of heartbeats stemmed from an issue where the shard got ratelimited, incorrectly attributed to the shard receiving `Heartbeat`s (in reality, as this PR shows, it was ratelimited due to the ratelimiter algorithm being broken). By reserving an additional command I'm already keeping things conservative and there's no logic for why specifically double the amount of heartbeats is correct, one could as easily say it should be triple or even more. I have successfully gotten ratelimited with the old implementation, even when it reserved 2x the amounts of heartbeats, but I've never hit a ratelimit with the new one reserving one only additional heartbeat.
One can look at other libraries and notice that, except for those copying the behavior of Twilight, no other reserves these many commands per reset.

Reply:
Regarding reserving double the amount of heartbeats, I'll decide to go with your opinion. Now that the rate limiter is correct, that sounds fine to me.

Regarding the style of the function, as I see it, there's two separate things that I'm noticing: ignoring clippy lints and readability. For the former, I would be willing to follow you and say we should just ignore the lints, since their names are pretty descriptive of their behavior. 

For the latter, however, I'd like this function in particular, since it is doing math,  to become its own proof by showing the steps of the calculation. Any amount of golfing the code down is going to make the function seem more snappy, but it all compiles down to pretty much the same thing.

Reply:
I'm not quite following, could you exemplify how you want the function to look like?